services/repository/contributors_graph.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	repo_model "code.gitea.io/gitea/models/repo"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	api "code.gitea.io/gitea/modules/structs"
+	util "code.gitea.io/gitea/modules/util"
+)
+
+// CreateWeeks converts list of sundays to list of *api.WeekData
+func CreateWeeks(sundays []int64) []*api.WeekData {
+	var weeks []*api.WeekData
+	for _, week := range sundays {
+		weeks = append(weeks, &api.WeekData{
+			Week:      week,
+			Additions: 0,
+			Deletions: 0,
+			Commits:   0,
+		},
+		)
+	}
+	return weeks
+}
+
+// GetContributorStats returns contributors stats for git commits for given revision or default branch
+func GetContributorStats(ctx context.Context, repo *repo_model.Repository, revision string) (map[string]*api.ContributorData, error) {
+	gitRepo, closer, err := git.RepositoryFromContextOrOpen(ctx, repo.RepoPath())
+	if err != nil {
+		return nil, fmt.Errorf("OpenRepository: %w", err)
+	}
+	defer closer.Close()
+
+	if len(revision) == 0 {
+		revision = repo.DefaultBranch
+	}
+	extendedCommitStats, err := gitRepo.ExtendedCommitStats(revision)
+	if err != nil {
+		return nil, fmt.Errorf("ExtendedCommitStats: %w", err)
+	}
+
+	layout := "2006-01-02"
+	initialCommitDate := extendedCommitStats[0].Author.Date
+
+	startingSunday, _ := util.FindLastSundayBeforeDate(initialCommitDate)
+	endingSunday, _ := util.FindFirstSundayAfterDate(time.Now().Format(layout))
+
+	sundays, _ := util.ListSundaysBetween(startingSunday, endingSunday)
+
+	unknownUserAvatarLink := user_model.NewGhostUser().AvatarLink(ctx)
+	contributorsCommitStats := make(map[string]*api.ContributorData)
+	contributorsCommitStats["total"] = &api.ContributorData{
+		Name:       "Total",
+		AvatarLink: unknownUserAvatarLink,
+		Weeks:      CreateWeeks(sundays),
+	}
+	total := contributorsCommitStats["total"]
+
+	for _, v := range extendedCommitStats {
+		if len(v.Author.Email) == 0 {
+			continue
+		}
+		if _, ok := contributorsCommitStats[v.Author.Email]; !ok {
+			u, err := user_model.GetUserByEmail(ctx, v.Author.Email)
+			if u == nil || user_model.IsErrUserNotExist(err) {
+				contributorsCommitStats[v.Author.Email] = &api.ContributorData{
+					Name:       v.Author.Name,
+					AvatarLink: unknownUserAvatarLink,
+					Weeks:      CreateWeeks(sundays),
+				}
+			} else {
```

Comment:
error case missing.
Do you perhaps mean
```suggestion
			} else if err == nil {
```
?
And how do we handle the case `err != nil`?

Reply:
Yeah, you are right. I would normally handle all the cases but my knowledge in go is limited. So I don't know what could go wrong and how should I handle that case. Maybe @silverwind could help me resolving this.

Reply:
You are only checking `IsErrUserNotExist`, but in theory it could be another error as well, so I guess the suggestion for `err == nil` is likely fine to prevent a runtime crash when `u` is nil.

Reply:
> so I guess the suggestion for `err == nil` is likely fine to prevent a runtime crash when `u` is nil.

I think you meant "when `u` is **not** nil and there is an error other than `IsErrUserNotExist`". Because when `u` is nil, it will fall into first if statement already. 

I gave this a second thought and I think I can safely delete `user_model.IsErrUserNotExist(err)` from the first if statement. Because if there is an error, `u` will be nil. I don't care what the error is, I just create a new entry for that user. If `u` is not nil, that means we get our user so we will continue with the `else` part.

This assumes that the functions I am calling follows the conventions you described (i.e. either a result and nil error or a nil result and error). If this is the case, then there is no need to check `err` in first if statement.

Reply:
I have removed error checking. Let me know what you think: https://github.com/go-gitea/gitea/pull/25439/commits/f2d33949c342ce7f2946d18a60b0964f56312119